protocols/purple/purple.c

```diff
@@ -1055,9 +1068,57 @@ static void prplcb_request_test()
 }
 */
 
+struct request_input_data {
```

Comment:
move this to the top!
also move that callback_t typedef outside of purple_request_input_callback
and use callback_t for the data_callback type instead of GCallback, save you some casting when calling it.